src/tests/zf_apps/zfsink.c

```diff
@@ -243,14 +245,80 @@ void print_attrs(struct zf_attr* attr)
 }
 
 
-static void monitor()
+static int zf_stats_header_print(struct zf_stack* stack, int layout_sz,
+                                  const zf_stats_layout** zf_stats_layout)
+{
+  int i, nic_cnt;
+  nic_cnt = zf_stats_query_layout(stack, zf_stats_layout, layout_sz);
+
+  /* only display names for the first NIC and assume others are the same */
+  if( nic_cnt > 1 ) {
+    printf("\n#");
+  }
+  for( i = 0; i < zf_stats_layout[0]->evsl_fields_num; ++i )
+    printf("  %10s", zf_stats_layout[0]->evsl_fields[i].evsfl_name);
+
+  return nic_cnt;
+}
+
+
+static void zf_stats_print(struct zf_stack* stack, int reset_stats,
+                           uint8_t* stats_data, int nic_cnt,
+                           const zf_stats_layout** zf_stats_layout)
+{
+  int i, n, n_pad;
+  uint8_t* cur_data = stats_data;
+
+  zf_stats_query(stack, stats_data, zf_stats_layout, nic_cnt, reset_stats);
+
+  for( n = 0; n < nic_cnt; n++ ) {
+    if( nic_cnt > 1 )
+      printf("\n");
+    for( i = 0; i < zf_stats_layout[n]->evsl_fields_num; ++i ) {
+      const zf_stats_field_layout* f = &zf_stats_layout[n]->evsl_fields[i];
+      n_pad = strlen(f->evsfl_name);
+      if( n_pad < 10 )
+        n_pad = 10;
+      switch( f->evsfl_size ) {
+        case sizeof(uint32_t):
+          printf("  %*d", n_pad, *(uint32_t*)(cur_data + f->evsfl_offset));
+          break;
+        default:
+          printf("  %*s", n_pad, ".");
+      };
+    }
+    cur_data += zf_stats_layout[n]->evsl_data_size;
+  }
+}
+
+
+static void monitor(struct zf_stack* stack)
 {
   uint64_t now_bytes, prev_bytes;
   struct timeval start, end;
   uint64_t prev_pkts, now_pkts;
   int ms, pkt_rate, mbps;
 
-  vlog("#%9s %16s %16s", "pkt-rate", "bandwidth(Mbps)", "total-pkts\n");
+  const zf_stats_layout** zf_stats_layout;
+  uint8_t* stats_data = NULL;
+  int i, nic_cnt = 0, malloc_sz = 0;
+
+  nic_cnt = zf_stack_num_intfs(stack);
+  zf_stats_layout = malloc(nic_cnt * sizeof(struct zf_stats_layout *));
+  ZF_TEST(zf_stats_layout != NULL);
+
+  pthread_mutex_lock(&printf_mutex);
+  printf("#%9s %16s %16s", "pkt-rate", "bandwidth(Mbps)", "total-pkts");
+
+  if( cfg_intf_stats ) {
+    nic_cnt = zf_stats_header_print(stack, nic_cnt, zf_stats_layout);
+    for( i = 0; i < nic_cnt; i++ )
+      malloc_sz += zf_stats_layout[i]->evsl_data_size;
```

Comment:
rather than making the user do this, I'd have a utility to allocate data sufficient for new data structure containing layouts mentioned above